### PR TITLE
[Test] Cypress static code analysis test

### PR DESCRIPTION
**Pull request by jhainzlschmid**
### Motivation and Context

We want to have an E2E test to make sure that static code analysis works for a student participating in a programming exercise.

### Description

This PR adds another programming exercise test, which creates a programming exercise with static code analysis enabled. SCA can reduce up to 50% (5 points) . The test imitates a student and enters code, which passes all tests of the exercises, but contains some SCA issues. It then makes sure that all SCA issues are shown in the feedback to the student and that it is transparent which issue caused how many points to be reduced.

### Steps for Testing

Tests are executed by [bamboo](https://bamboo.ase.in.tum.de/browse/ARTEMIS-AETBBT-QE-222), so only code reviews are required

### Review Progress

- Code Review
  - [ ] Review 1
  - [ ] Review 2

### CODE REVIEW

**Comment by jhainzlschmid** [comment]
[Tests still pass](https://bamboo.ase.in.tum.de/browse/ARTEMIS-AETBBT-236) after resolving merge conflict

**Review by rs-maker — APPROVED** [review]
Good one cant complain :+1:

File: src/test/cypress/integration/ProgrammingExerciseStaticCodeAnalysis.spec.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+     * Creates a course and a programming exercise inside that course.
+     */
+    function setupCourseAndProgrammingExercise() {
+        cy.login(users.getAdmin());
+        courseManagement.createCourse().then((response) => {
+            course = response.body;
+            courseManagement.addStudentToCourse(course.id, users.getStudentOne().username);
+            courseManagement
+                .createProgrammingExercise({ course }, 50)
+                .its('body')
```
Comment (julian-christl) [inline src/test/cypress/integration/ProgrammingExerciseStaticCodeAnalysis.spec.ts:53]: Since you replaced it in another file I'd suggest moving the body down in the then

**Review by julian-christl — CHANGES_REQUESTED** [review]
Looks good, have some questions.

File: src/test/cypress/support/pageobjects/exercises/programming/ScaFeedbackModal.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    }
+
+    shouldShowCodeIssue(feedbackText: string, pointReduction: string) {
+        cy.get('.feedback-text')
+            .contains(feedbackText)
+            .scrollIntoView()
+            .should('be.visible')
+            .parents('.feedback-item')
+            .find('.feedback-points')
+            .scrollIntoView()
```
[the remark this answers is not in the document]
Comment (jhainzlschmid) [reply src/test/cypress/support/pageobjects/exercises/programming/ScaFeedbackModal.ts:23]: Those are actually multiple objects. First we search for the feedback text (red), scroll it into view and assert that it is visible. Then we get the feedback points (blue) and do the same here. Cypress scrolls the element into view that it is basically at the edge of the screen. I have had it happen that if I only scroll one of the elements into view the other one is not. So as safety measure I added it for both elements

File: src/test/cypress/support/requests/CourseManagementRequests.ts [hunk 3]
```diff
@@ -23,6 +24,8 @@ export class CourseManagementRequests {
      * @returns <Chainable> request response
      */
     deleteCourse(id: number) {
+        // Sometimes the backend fails with a ConstraintViolationError if we delete the course immediately
+        cy.wait(1000);
```
[the remark this answers is not in the document]
Comment (jhainzlschmid) [reply src/test/cypress/support/requests/CourseManagementRequests.ts:28]: I just wanted to be on the safe side. I will reduce it to 100ms and see how it goes

File: src/test/cypress/support/pageobjects/OnlineEditorPage.ts [hunk 4]
```diff
@@ -126,6 +131,39 @@ export class OnlineEditorPage {
...
+    cy.log('Participating in the programming exercise as a student...');
+    cy.contains(courseName).parents('.card-header').click();
+    cy.url().should('include', '/exercises');
+    cy.intercept(POST, COURSE_BASE + '*/exercises/*/participations').as('participateInExerciseQuery');
+    cy.get(exerciseRow).contains(programmingExerciseName).should('be.visible');
+    cy.get(exerciseRow).find('.start-exercise').click();
+    cy.wait('@participateInExerciseQuery');
+    cy.intercept(GET, BASE_API + 'programming-exercise-participations/*/student-participation-with-latest-result-and-feedbacks').as('initialQuery');
+    cy.get(exerciseRow).find('[buttonicon="folder-open"]').click();
+    cy.wait('@initialQuery').wait(2000);
```
[the remark this answers is not in the document]
Comment (jhainzlschmid) [reply src/test/cypress/support/pageobjects/OnlineEditorPage.ts:164]: I just copied the function from over tbh. I think the 2s wait isn't necessary anymore because we increased the default timeout of Cypress in the past. I will remove it and see how it goes

[the remark this answers is not in the document]
Comment (julian-christl) [reply src/test/cypress/support/pageobjects/OnlineEditorPage.ts:164]: Code clone alert? 👀

File: src/test/cypress/support/pageobjects/OnlineEditorPage.ts [hunk 5]
```diff
@@ -126,6 +131,39 @@ export class OnlineEditorPage {
     }
 }
 
+/**
+ * General method for entering, submitting and verifying something in the online editor.
+ */
+export function makeSubmissionAndVerifyResults(editorPage: OnlineEditorPage, packageName: string, submission: ProgrammingExerciseSubmission, verifyOutput: () => void) {
+    // We create an empty file so that the file browser does not create an extra subfolder when all files are deleted
+    editorPage.createFileInRootPackage('placeholderFile');
```
[the remark this answers is not in the document]
Comment (jhainzlschmid) [reply src/test/cypress/support/pageobjects/OnlineEditorPage.ts:139]: We could remove it, but it doesn't affect the compilation in any way, so I don't see the value of this. In the best case nothing changes and in the worst case I add some flakiness into the test

File: src/test/cypress/integration/ProgrammingExerciseStaticCodeAnalysis.spec.ts (hunk 1 above)
Comment (jhainzlschmid) [reply src/test/cypress/integration/ProgrammingExerciseStaticCodeAnalysis.spec.ts:53]: True, I think I am not really consistent with that. I will move it into the then block

File: src/test/cypress/integration/ProgrammingExerciseParticipation.spec.ts [hunk 6]
```diff
@@ -61,13 +53,9 @@ describe('Programming exercise participations', () => {
...
             courseManagement.addStudentToCourse(course.id, users.getStudentThree().username);
-            courseManagement
-                .createProgrammingExercise({ course })
-                .its('body')
-                .then((exercise) => {
-                    expect(exercise).to.not.be.null;
-                    programmingExercise = exercise;
-                });
+            courseManagement.createProgrammingExercise({ course }).then((exerciseResponse) => {
+                exercise = exerciseResponse.body;
```
[the remark this answers is not in the document]
Comment (jhainzlschmid) [reply src/test/cypress/integration/ProgrammingExerciseParticipation.spec.ts:57]: I don't see the value of the assertion here. The backend shouldn't answer with null in the first place if the request was successfully executed. If it unexpectedly returns null or no body the test will fail anyways because it needs the exercise object. If I add the assertion again it should probably be `to.be.ok` to cover the theoretical possibility that the backend returns null as body

File: src/test/cypress/support/pageobjects/OnlineEditorPage.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/test/cypress/support/pageobjects/OnlineEditorPage.ts:139]: I agree, would it make more sense to not use the file approach but recreate the folder?
Because these e2e tests should test the way a user would do it, correct? And the user would probably delete the files which lead to the need to apparently recreate the folder.

While it doesn't change anything right now, it doesn't represent the way a user would do it and could therefore capture in the future only certain but not all parts.
Or do I overthink this here?

File: src/test/cypress/integration/ProgrammingExerciseStaticCodeAnalysis.spec.ts [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        cy.login(users.getAdmin());
+        scaConfig.visit(course.id, exercise.id);
+        scaConfig.makeEveryScaCategoryInfluenceGrading();
+        scaConfig.saveChanges();
+    });
+
+    it('Makes successful submission with SCA errors', function () {
+        startParticipationInProgrammingExercise(course.title, exercise.title, users.getStudentOne());
+        makeSuccessfulSubmissionWithScaErrors();
+    });
```
[the remark this answers is not in the document]
Comment (jhainzlschmid) [reply src/test/cypress/integration/ProgrammingExerciseStaticCodeAnalysis.spec.ts]: Yeah the idea was that the first test adjust the SCA configuration as tutor and then the second test participates as a student and verifies that he gets the SCA feedback and reduced points etc. I agree that it is not great to have tests depending on each other, but there aren't too many options we have. We can either merge all tests, which depend on each other, into one large test or we repeat the setup for every test. One huge test isn't great, because it's not really readable (in this case it would be, but look at the ModelingExercise.spec for example). Repeating the setup  for every sub-test takes a lot of implementation and execution resources.
So I went with this structure, because the tests clearly separate what is tested and are readable

File: src/test/cypress/support/pageobjects/OnlineEditorPage.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (jhainzlschmid) [reply src/test/cypress/support/pageobjects/OnlineEditorPage.ts:164]: Copied was the wrong word. I moved it, so the function can be shared between spec files :p

File: src/test/cypress/support/pageobjects/OnlineEditorPage.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (jhainzlschmid) [reply src/test/cypress/support/pageobjects/OnlineEditorPage.ts:139]: My dream solution would be that no files is deleted at all and that the E2E test behaves like the user, but that is simply not possible in this scenario. A real user would open the programming exericse, open a file, and then insert his code into specific locations of the code template (like filling out the sort method etc). Cypress can't do this in a reliable way, so my previous solution was that when Cypress opens a file it deletes the content of the file by sending CTRL + A + DEL. That solution added a lot of flakiness. Sometimes cypress failed to execute that exact keypress combination in the right time and then the content was not deleted, which resulted in a build failure and as such a test failure.
That's why I moved to Cypress first deleting the file (to get rid of the content) and then creating a new file with the exact same name and writes the content in there. This is a lot more stable, but yeah as you mentioned it's not really what a user would do. Overall I still think that the test is very valuable in what it actually does test (file deletion, file creation, entering code via the ace code editor and checking the results)

File: src/test/cypress/integration/ProgrammingExerciseStaticCodeAnalysis.spec.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/test/cypress/integration/ProgrammingExerciseStaticCodeAnalysis.spec.ts]: I see. The issue I have with it is that it relies on the specific execution order and that the tests are run in the same thread. In case we want to parallelise this in any way or make it more flexible, issues would arise.

Also from looking at a test case, it is not clear that they are depending on each other => Moving something could get messy.

I agree that having blob tests is not a solution and that implementation and execution resources are issues to tackle.

The first two are solvable by extracting certain parts into functions that get called by tests. This also reduces code duplication and you already do it. So I don't see the issue of applying this here as well if the case arises more than just this time.

Execution resources could be an issue but I can't see it in this specific case. Do relate to other test cases where this issue would arise?
If there is a real issue with that, one could extract all dependant tests into a test file and specify the setup in the beforeEach() method.

What do you think?

File: src/test/cypress/support/pageobjects/OnlineEditorPage.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/test/cypress/support/pageobjects/OnlineEditorPage.ts:164]: haha, okay then let's see how it will work out.

File: src/test/cypress/support/pageobjects/OnlineEditorPage.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/test/cypress/support/pageobjects/OnlineEditorPage.ts:139]: I see, thanks for the explanation. Good to know that there is thought behind this decision.
Though, I never heard of the combination of CTRL + A + DEL. What I know is CTRL + A to mark everything and then DEL to delete it. If you tried to execute the tripel, could it be the case that sometimes DEL was executed before A and therefore nothing was deleted?
Depending on a) whether you already investigated this or not and b) this only affects files in this PR or also other files, I would solve this in this PR, solve it in another PR or just ignore it then.

[the remark this answers is not in the document]
Comment (jhainzlschmid) [reply src/test/cypress/support/pageobjects/OnlineEditorPage.ts:139]: Yeah I spent quite some time trying to get it stable. It sent CTRL + A and DEL with a delay (I also tried different delays). All options I tried with this approach (trying to clear the text via keyboard input) were flaky, so I went for the file deletion recreation approach.
I can have another look in the future, but I wouldn't change the approach now in this PR.

File: src/test/cypress/integration/ProgrammingExerciseStaticCodeAnalysis.spec.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (jhainzlschmid) [reply src/test/cypress/integration/ProgrammingExerciseStaticCodeAnalysis.spec.ts]: Yeah, I can put it into the same test and refactor the individual steps into their own high-level methods. With execution resources I mostly meant how long the test is running. We need to wait one minute between creating a course and creating a programming exercise. Duplicating the setup for each tests would mean that we probably add about 90 seconds of time to the test for nothing. This could still be avoided if we create only one course and then two programming exercises, but the programming exercise creation itself creates quite some time, so we would probably still add about 20 seconds of execution time.
Furthermore the tests then kinda miss the correlation from the UI interaction in the first test (configuring the SCA grading)  with the effect it has on the second test (that all SCA categories are enabled and reduce points).
So in the first test we would test that we can change the grading configuration in the UI (but don't assert if it actually has an impact on the student exercise).
Moreover I would have to implement an additional request for the second test (so a programming exercise is created, cypress then configures SCA via this request and then the student can participate in the exercise). This request will probably not be used anywhere else and needs to be maintained.
So yeah I prefer moving the tests into one larger test in this case.
If we add parallelisation at some point via the Cypress Dashboard it will[ only parallelize complete spec files](https://docs.cypress.io/guides/guides/parallelization#Overview) not the individual tests inside a spec file.
image

File: src/test/cypress/support/pageobjects/OnlineEditorPage.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/test/cypress/support/pageobjects/OnlineEditorPage.ts:139]: Is there a reason why you didn't use the cypress clear function? https://docs.cypress.io/api/commands/clear
Sorry for the naive question. Just learning about cypress so... :D

File: src/test/cypress/integration/ProgrammingExerciseStaticCodeAnalysis.spec.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/test/cypress/integration/ProgrammingExerciseStaticCodeAnalysis.spec.ts]: Putting it in one test sounds good. Especially since you can split it into multiple (non-test-)functions. But then it is for a new face easy to understand I think.

File: src/test/cypress/support/pageobjects/OnlineEditorPage.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (jhainzlschmid) [reply src/test/cypress/support/pageobjects/OnlineEditorPage.ts:139]: Yeah I tried that with cy.clear(). I also tried to use cy.invoke('text', ''). But both did not work. Cypress complains that the element is not clearable. To be honest I think we select the wrong element to type into. I found a [somewhat related thread](https://github.com/cypress-io/cypress/issues/1818). I will try the selector that the cypress developer proposed and see if that one is clearable.

[the remark this answers is not in the document]
Comment (jhainzlschmid) [reply src/test/cypress/support/pageobjects/OnlineEditorPage.ts:139]: So I managed to clear the field with the selector proposed by the cypress developer, but it fails after the writing into the first file. The issue seems to be that multiple 'ace_text-input' elements exist after switching to another file. I will investigate this further in a followup PR since this also affects other specs (ProgrammingExerciseParticipation and ExamParticipation).

**Comment by jhainzlschmid** [comment]
I added some suggested improvements and merged develop. The tests [still pass](https://bamboo.ase.in.tum.de/browse/ARTEMIS-AETBBT-270)

File: src/test/cypress/support/requests/CourseManagementRequests.ts [hunk 8]
```diff
@@ -13,6 +13,8 @@ import quizTemplate from '../../fixtures/quiz_exercise_fixtures/quizExercise_tem
 export const COURSE_BASE = BASE_API + 'courses/';
 export const COURSE_MANAGEMENT_BASE = BASE_API + 'course-management/';
 export const PROGRAMMING_EXERCISE_BASE = BASE_API + 'programming-exercises/';
+const QUIZ_EXERCISE_BASE = BASE_API + 'quiz-exercises/';
```
Comment (julian-christl) [inline src/test/cypress/support/requests/CourseManagementRequests.ts:16]: Is there a reason why you don't export this? Even if you don't use it somewhere else I think in this context it makes sense to export it anyway to stay uniform

File: src/test/cypress/support/pageobjects/OnlineEditorPage.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (julian-christl) [reply src/test/cypress/support/pageobjects/OnlineEditorPage.ts:139]: Sound great. But amazing that we could make some progress don't that issue 👍

File: src/test/cypress/support/pageobjects/ArtemisPageobjects.ts [hunk 9]
```diff
@@ -26,14 +31,23 @@ export class ArtemisPageobjects {
     navigationBar = new NavigationBar();
     examCreation = new ExamCreationPage();
     examManagement = new ExamManagementPage();
+    examStartEnd = new ExamStartEndPage();
+    examNavigationBar = new ExamNavigationBar();
+    programmingExerciseCreation = new ProgrammingExerciseCreationPage();
```
Comment (julian-christl) [inline src/test/cypress/support/pageobjects/ArtemisPageobjects.ts]: Why was this added? Isn't this a duplicate of `programmingExercise.creation`?

Comment (jhainzlschmid) [reply src/test/cypress/support/pageobjects/ArtemisPageobjects.ts]: Good catch. Yeah its a duplicate. I removed it

File: src/test/cypress/support/requests/CourseManagementRequests.ts (hunk 8 above)
Comment (jhainzlschmid) [reply src/test/cypress/support/requests/CourseManagementRequests.ts]: Yeah it was added by the merge from develop. I agree, it makes sense to also export that constant

**Review by julian-christl — APPROVED** [review]
Thanks for all the changes and discussions!
The code looks good to me now.

**Comment by jhainzlschmid** [comment]
[Tests still pass](https://bamboo.ase.in.tum.de/browse/ARTEMIS-AETBBT-271) after the changes

**Review by rs-maker — APPROVED** [review]
I still agree with this.
